README.md
# NAME

Regexp::VerbalExpressions - Perl Regular expressions made easy

# SYNOPSIS

    use Regexp::VerbalExpressions;

    # Create an example of how to test for correctly formed URLs
    my $re = verex
           ->start_of_line
           ->then('http')
           ->maybe('s')
           ->then('://')
           ->maybe('www.')
           ->anything_but(' ')
           ->end_of_line;

    if ('https://www.google.com/' =~ $re) {
        print 'We have a correct URL'; # This output will fire
    } else {
        print 'The URL is incorrect';
    }

    print $re; # Outputs the actual expression used: ^(?:http)(?:s)?(?:\:\/\/)(?:www\.)?(?:[^\ ]*)$

# DESCRIPTION

Regexp::VerbalExpressions is a Perl module that helps to construct difficult regular expressions.

# API

## Modifiers

- `anything()`
- `anything_but($value)`
- `end_of_line()`
- `find($value)`
- `maybe($value)`
- `start_of_line()`
- `then($value)`

## Special characters and groups

- `any($value)`
- `any_of($value)`
- `br()`
- `line_break()`
- `range($from, $to)`
- `tab()`
- `word()`

## Modifiers

- `with_any_case()`
- `stop_at_first()`
- `search_one_line()`

## Functions

- `replace($source, $value)`

## Other

- `add($expression)`
- `multiple($value)`
- `or()`

## Other implementations  
You can view all implementations on [VerbalExpressions.github.io](http://VerbalExpressions.github.io)

# LICENSE

Copyright (C) Takumi Akiyama.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Takumi Akiyama <t.akiym@gmail.com>


Changes
Revision history for Perl extension Regexp-VerbalExpressions

{{$NEXT}}

0.02 2013-08-10T02:25:58Z
    - fix for perl-5.13.6 or less
    - add something(), something_but()

0.01 2013-08-09T10:10:21Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Regexp-VerbalExpressions',
    module_name     => 'Regexp::VerbalExpressions',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Regexp/VerbalExpression.pm
package Regexp::VerbalExpression;
use 5.008005;
use strict;
use warnings;
use parent qw/Exporter/;
use overload fallback => 1, q{""} => \&source, 'qr' => \&regex;

our $VERSION = "0.01";

our @EXPORT = qw/verex/;

sub new {
    my $class = shift;
    return bless {
        s         => '',
        modifiers => {
            i => '',
            g => '',
            m => '',
        },
    }, $class;
}

sub verex { __PACKAGE__->new() }

sub add {
    my ($self, $value) = @_;
    $self->{s} .= $value;
    return $self;
}

sub regex {
    my $self = shift;
    my $modifiers = join '', values %{$self->{modifiers}};
    return qr/(?^$modifiers:$self->{s})/;
}
*compile = \&regex;

sub source {
    my $self = shift;
    return $self->{s};
}
*raw = *value = \&source;

sub find {
    my ($self, $value) = @_;
    return $self->add('(?:' . quotemeta($value) . ')');
}
*then = \&find;

sub start_of_line {
    my $self = shift;
    return $self->add('^');
}

sub end_of_line {
    my $self = shift;
    return $self->add('$');
}

sub maybe {
    my ($self, $value) = @_;
    return $self->add('(?:' . quotemeta($value) . ')?');
}

sub anything {
    my $self = shift;
    return $self->add('(?:.*)');
}

sub anything_but {
    my ($self, $value) = @_;
    return $self->add('(?:[^' . quotemeta($value) . ']*)');
}

sub line_break {
    my $self = shift;
    return $self->add('(?:\n|(?:\r\n))');
}
*br = \&line_break;

sub tab {
    my $self = shift;
    return $self->add('\t');
}

sub word {
    my $self = shift;
    return $self->add('\w+');
}

sub any {
    my ($self, $value) = @_;
    return $self->add('[' . quotemeta($value) . ']');
}
*any_of = \&any;

sub range {
    my ($self, @args) = @_;
    my $value = '[';
    for (my $i = 0; $i < @args; $i += 2) {
        $value .= quotemeta($args[$i]) . '-' . quotemeta($args[$i + 1]);
    }
    $value .= ']';
    return $self->add($value);
}

sub multiple {
    my ($self, $value) = @_;
    $self->add(quotemeta($value) . '+');
}

sub or {
    my ($self, $value) = @_;
    $self->add('|');
    return defined $value ? $self->find($value) : $self;
}
*alternatively = \&or;

sub replace {
    my ($self, $source, $value) = @_;
    if ($self->{modifiers}{g}) {
        local $self->{modifiers}{g} = ''; # not include /g in pattern
        my $re = $self->regex;
        $source =~ s/$re/$value/g;
    } else {
        my $re = $self->regex;
        $source =~ s/$re/$value/;
    }
    return $source;
}

sub with_any_case {
    my ($self, $value) = @_;
    $self->{modifiers}{i} = $value ? 'i' : '';
    return $self;
}

sub stop_at_first {
    my ($self, $value) = @_;
    $self->{modifiers}{g} = $value ? 'g' : '';
    return $self;
}

sub search_one_line {
    my ($self, $value) = @_;
    $self->{modifiers}{m} = $value ? 'm' : '';
    return $self;
}

1;
__END__

=encoding utf-8

=head1 NAME

Regexp::VerbalExpression - Perl Regular expressions made easy

=head1 SYNOPSIS

    use Regexp::VerbalExpression;

    # Create an example of how to test for correctly formed URLs
    my $re = verex
           ->start_of_line
           ->then('http')
           ->maybe('s')
           ->then('://')
           ->maybe('www.')
           ->anything_but(' ')
           ->end_of_line;

    if ('https://www.google.com/' =~ $re) {
        print 'We have a correct URL'; # This output will fire
    } else {
        print 'The URL is incorrect';
    }

    print $re; # Outputs the actual expression used: ^(?:http)(?:s)?(?:\:\/\/)(?:www\.)?(?:[^\ ]*)$

=head1 DESCRIPTION

Regexp::VerbalExpression is a Perl module that helps to construct difficult regular expressions.

=head1 API

=head2 Modifiers

=over 4

=item C<< anything() >>

=item C<< anything_but($value) >>

=item C<< end_of_line() >>

=item C<< find($value) >>

=item C<< maybe($value) >>

=item C<< start_of_line() >>

=item C<< then($value) >>

=back

=head2 Special characters and groups

=over 4

=item C<< any($value) >>

=item C<< any_of($value) >>

=item C<< br() >>

=item C<< line_break() >>

=item C<< range($from, $to) >>

=item C<< tab() >>

=item C<< word() >>

=back

=head2 Modifiers

=over 4

=item C<< with_any_case() >>

=item C<< stop_at_first() >>

=item C<< search_one_line() >>

=back

=head2 Functions

=over 4

=item C<< replace($source, $value) >>

=back

=head2 Other

=over 4

=item C<< add($expression) >>

=item C<< multiple($value) >>

=item C<< or() >>

=back

=head1 SEE ALSO

L<https://github.com/VerbalExpressions/JSVerbalExpressions>

=head1 LICENSE

Copyright (C) Takumi Akiyama.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Takumi Akiyama E<lt>t.akiym@gmail.comE<gt>

=cut



lib/Regexp/VerbalExpressions.pm
package Regexp::VerbalExpressions;
use 5.008005;
use strict;
use warnings;
use parent qw/Exporter/;
use overload fallback => 1, q{""} => \&source, 'qr' => \&regex;

our $VERSION = "0.02";

our @EXPORT = qw/verex/;

sub new {
    my $class = shift;
    return bless {
        s         => '',
        modifiers => {
            i => 0,
            m => 0,
            s => 0,
            x => 0,
        },
        global_matching => 0,
    }, $class;
}

sub verex { __PACKAGE__->new() }

sub add {
    my ($self, $value) = @_;
    $self->{s} .= $value;
    return $self;
}

sub regex {
    my $self = shift;
    my $enabled = my $disabled = '';
    for my $modifier (sort keys %{$self->{modifiers}}) {
        if ($self->{modifiers}{$modifier}) {
            $enabled .= $modifier;
        } else {
            $disabled .= $modifier;
        }
    }
    return qr/(?$enabled-$disabled:$self->{s})/;
}
*compile = \&regex;

sub source {
    my $self = shift;
    return $self->{s};
}
*raw = *value = \&source;

sub find {
    my ($self, $value) = @_;
    return $self->add('(?:' . quotemeta($value) . ')');
}
*then = \&find;

sub start_of_line {
    my $self = shift;
    return $self->add('^');
}

sub end_of_line {
    my $self = shift;
    return $self->add('$');
}

sub maybe {
    my ($self, $value) = @_;
    return $self->add('(?:' . quotemeta($value) . ')?');
}

sub anything {
    my $self = shift;
    return $self->add('(?:.*)');
}

sub anything_but {
    my ($self, $value) = @_;
    return $self->add('(?:[^' . quotemeta($value) . ']*)');
}

sub something {
    my $self = shift;
    return $self->add('(?:.+)');
}

sub something_but {
    my ($self, $value) = @_;
    return $self->add('(?:[^' . quotemeta($value) . ']+)');
}

sub line_break {
    my $self = shift;
    return $self->add('(?:\n|(?:\r\n))');
}
*br = \&line_break;

sub tab {
    my $self = shift;
    return $self->add('\t');
}

sub word {
    my $self = shift;
    return $self->add('\w+');
}

sub any {
    my ($self, $value) = @_;
    return $self->add('[' . quotemeta($value) . ']');
}
*any_of = \&any;

sub range {
    my ($self, @args) = @_;
    my $value = '[';
    for (my $i = 0; $i < @args; $i += 2) {
        $value .= quotemeta($args[$i]) . '-' . quotemeta($args[$i + 1]);
    }
    $value .= ']';
    return $self->add($value);
}

sub multiple {
    my ($self, $value) = @_;
    $self->add(quotemeta($value) . '+');
}

sub or {
    my ($self, $value) = @_;
    $self->add('|');
    return defined $value ? $self->find($value) : $self;
}
*alternatively = \&or;

sub replace {
    my ($self, $source, $value) = @_;
    my $re = $self->regex;
    if ($self->{global_matching}) {
        $source =~ s/$re/$value/g;
    } else {
        $source =~ s/$re/$value/;
    }
    return $source;
}

sub _build_modifier {
    my ($self, $modifier, $value) = @_;
    if (!defined $value) {
        # XXX verex->with_any_case;
        $self->{modifiers}{$modifier} = 1;
    } else {
        $self->{modifiers}{$modifier} = $value ? 1 : 0;
    }
}

sub with_any_case {
    my ($self, $value) = @_;
    $self->_build_modifier(i => $value);
    return $self;
}

sub stop_at_first {
    my ($self, $value) = @_;
    if (!defined $value) {
        $self->{global_matching} = 1;
    } else {
        $self->{global_matching} = $value ? 1 : 0;
    }
    return $self;
}

sub search_one_line {
    my ($self, $value) = @_;
    $self->_build_modifier(m => $value);
    return $self;
}

1;
__END__

=encoding utf-8

=head1 NAME

Regexp::VerbalExpressions - Perl Regular expressions made easy

=head1 SYNOPSIS

    use Regexp::VerbalExpressions;

    # Create an example of how to test for correctly formed URLs
    my $re = verex
           ->start_of_line
           ->then('http')
           ->maybe('s')
           ->then('://')
           ->maybe('www.')
           ->anything_but(' ')
           ->end_of_line;

    if ('https://www.google.com/' =~ $re) {
        print 'We have a correct URL'; # This output will fire
    } else {
        print 'The URL is incorrect';
    }

    print $re; # Outputs the actual expression used: ^(?:http)(?:s)?(?:\:\/\/)(?:www\.)?(?:[^\ ]*)$

=head1 DESCRIPTION

Regexp::VerbalExpressions is a Perl module that helps to construct difficult regular expressions.

=head1 API

=head2 Modifiers

=over 4

=item C<< anything() >>

=item C<< anything_but($value) >>

=item C<< end_of_line() >>

=item C<< find($value) >>

=item C<< maybe($value) >>

=item C<< start_of_line() >>

=item C<< then($value) >>

=back

=head2 Special characters and groups

=over 4

=item C<< any($value) >>

=item C<< any_of($value) >>

=item C<< br() >>

=item C<< line_break() >>

=item C<< range($from, $to) >>

=item C<< tab() >>

=item C<< word() >>

=back

=head2 Modifiers

=over 4

=item C<< with_any_case() >>

=item C<< stop_at_first() >>

=item C<< search_one_line() >>

=back

=head2 Functions

=over 4

=item C<< replace($source, $value) >>

=back

=head2 Other

=over 4

=item C<< add($expression) >>

=item C<< multiple($value) >>

=item C<< or() >>

=back

=head1 SEE ALSO

L<https://github.com/VerbalExpressions/JSVerbalExpressions>

=head1 LICENSE

Copyright (C) Takumi Akiyama.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Takumi Akiyama E<lt>t.akiym@gmail.comE<gt>

=cut

